### ISSUE TITLE
Feature request:  custom recognizer extensibility

### ISSUE THREAD

**Issue by michhar** [Microsoft/botbuilder-js#272]
A very powerful concept is having the capability of custom NLP model integration and some guidance on writing a custom recognizer as an extension.  This could be powerful for the data science community.

**Comment by Stevenic** [comment]
In v4 we don't really have a formalized `Recognizer` interface. Recognizers are just classes so you can create any class you want. I'm not sure what we would document to be honest. Can you create new recognizers? absolutely! But they can work any way you want and have any shape you want them to. Do you have thoughts for what advice we'd give in docs?

**Comment by vishwacsena** [comment]
Closing due to no activity

**Comment by szul** [comment]
@Stevenic @vishwacsena In searching through current and closed issues, as well as the source code, it doesn't look like much of the recognizer stuff outside of LUIS is coming over to SDK v4, correct?

For example, SDK v3 had `builder.RegExpRecognizer()`, and you could create a custom recognizer with `bot.recognizer()`.

Are there equivalents to these in SDK v4, or is the assumption that you can just create your own class/method to process the text from `context.activity.message` since all of that is handled with the `processActivity()` method (or it could be handled with a middleware component).

Apologies for commenting on a closed issue, but this seemed the most appropriate thread to get clarity rather than creating a new issue.

**Comment by stevengum** [comment]
> is the assumption that you can just create your own class/method to process the text from context.activity.message since all of that is handled with the processActivity() method (or it could be handled with a middleware component).

This is the right approach. We're not bringing the Recognizer stuff (see [interface](https://docs.botframework.com/en-us/node/builder/chat-reference/interfaces/_botbuilder_d_.iintentrecognizer.html)) over because there isn't a need for it. You don't "plug" your "recognizer" into the bot, you just call it when it's necessary. 

You might make a middleware that _caches_ a recognizer on `context.turnState` so that it can be called later in dialogs, or you might make a middleware that calls a NLP service and caches the _results_ on `context.turnState`. 

The latter is what we used to do with LUIS, but we decided to [take it out](https://github.com/Microsoft/botbuilder-js/pull/292) because calling LUIS is more complicated than just calling it for every incoming activity. In some cases, getting the results from LUIS was wasteful as it would just get dropped on the floor.
